crypto/compactcert/common.go

```diff
@@ -0,0 +1,88 @@
+// Copyright (C) 2019-2020 Algorand, Inc.
+// This file is part of go-algorand
+//
+// go-algorand is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package compactcert
+
+import (
+	"encoding/binary"
+	"fmt"
+	"math/big"
+
+	"github.com/algorand/go-algorand/crypto"
+)
+
+// hashCoin returns a number in [0, signedWeight) with a nearly uniform
+// distribution, "randomized" by all of the supplied arguments.
+func hashCoin(j uint64, sigcom crypto.Digest, signedWeight uint64) uint64 {
+	hashinput := make([]byte, 16+crypto.DigestSize)
+	binary.LittleEndian.PutUint64(hashinput[0:], j)
+	binary.LittleEndian.PutUint64(hashinput[8:], signedWeight)
+	copy(hashinput[16:], sigcom[:])
+	h := crypto.Hash(hashinput)
+
+	i := &big.Int{}
+	i.SetBytes(h[:])
+
+	w := &big.Int{}
+	w.SetUint64(signedWeight)
+
+	res := &big.Int{}
+	res.Mod(i, w)
```

Comment:
I'm still working on understanding this PR, but in general is taking `hash value % signedWeight` sufficiently uniformly random? I see your comment mentions 'nearly uniform' but its unclear to me what this distribution actually looks like.

Reply:
If I understand it correctly, it's retaining the randomness of the hashed value. i.e. it has the same entropy of hash(rand()), which might be less than just rand().

Reply:
My understanding is that, yes, this is sufficiently uniform given the scale of values we are considering (signedWeight being tiny compared to the size of the hash).